torch/distributed/algorithms/join.py

```diff
@@ -213,3 +270,55 @@ def _notify_procs_to_terminate(self):
         # and throws a `RuntimeError` in Python 3.7+ (PEP 479), so we just
         # raise a `RuntimeError` here
         raise RuntimeError(f"Rank {self._rank} exhausted all inputs.")
+
+    @staticmethod
+    def notify_join_context(joinable: _Joinable):
+        r"""
+        Notifies the join context manager that the calling process has not yet
+        joined; then, if ``throw_on_early_termination=True``, checks if uneven
+        inputs have been detected (i.e. if one process has already joined) and
+        throws an exception if so.
+
+        This method should be called from a :class:`_Joinable` object before
+        its per-iteration collective communications. For example, this should
+        be called at the beginning of the forward pass in DDP.
+
+        Only the first :class:`_Joinable` object passed into the context
+        manager performs the collective communications in this method, and
+        for the others, this method is vacuous.
+
+        Arguments:
+            joinable (_Joinable): the :class:`_Joinable` object calling this
+                method.
+
+        Returns:
+            An async work handle for the all-reduce meant to notify the context
```

Comment:
We are retiring PG async work handle. It might be better to return the future object from `Work.get_future()`.

Reply:
I did not make the change in this PR since `Reducer::set_forward_pass_work_handle()` expects the async work handle and not the future.

I filed an issue to track this: https://github.com/pytorch/pytorch/issues/61661.